### PR TITLE
planner: support BatchPointGet for partition table

### LINKED ISSUE #21847: Partition table didn't support BatchPointGet
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sh
create table t1 (c int primary key)
PARTITION BY RANGE (c) (
    PARTITION p0 VALUES LESS THAN (6),
    PARTITION p1 VALUES LESS THAN (11),
    PARTITION p2 VALUES LESS THAN (16),
    PARTITION p3 VALUES LESS THAN (21)
);
Query OK, 0 rows affected (0.08 sec)

mysql> explain select * from t1 where c in (1,10);
+-----------------------------+---------+-----------+------------------------+------------------------------------------------------+
| id                          | estRows | task      | access object          | operator info                                        |
+-----------------------------+---------+-----------+------------------------+------------------------------------------------------+
| PartitionUnion_9            | 4.00    | root      |                        |                                                      |
| ├─TableReader_11            | 2.00    | root      |                        | data:TableRangeScan_10                               |
| │ └─TableRangeScan_10       | 2.00    | cop[tikv] | table:t1, partition:p0 | range:[1,1], [10,10], keep order:false, stats:pseudo |
| └─TableReader_13            | 2.00    | root      |                        | data:TableRangeScan_12                               |
|   └─TableRangeScan_12       | 2.00    | cop[tikv] | table:t1, partition:p1 | range:[1,1], [10,10], keep order:false, stats:pseudo |
+-----------------------------+---------+-----------+------------------------+------------------------------------------------------+
5 rows in set (0.00 sec)
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

Should use BatchPointGet

### 3. What did you see instead (Required)

Use TableReaderExecutor instead 

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```sh
mysql> SELECT tidb_version()
    -> ;
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                                       |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-1836-g3942c411f-dirty
Edition: Community
Git Commit Hash: 3942c411f3e25b61ffe5eb0d93864f11c7f9f495
Git Branch: forbid_pointget_cross_dc
UTC Build Time: 2020-12-17 04:16:42
GoVersion: go1.15rc1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by xiaodong-ji**
### What problem does this PR solve?

Issue Number: close #21847

Problem Summary: Partition table didn't support BatchPointGet.

### What is changed and how it works?

What's Changed:

How it Works:

### Check List

Tests

- Manual test
```
create table t1 (c int primary key)
PARTITION BY RANGE (c) (
    PARTITION p0 VALUES LESS THAN (6),
    PARTITION p1 VALUES LESS THAN (11),
    PARTITION p2 VALUES LESS THAN (16),
    PARTITION p3 VALUES LESS THAN (21)
);

mysql> explain select * from t1 where c in (1,10);
+--------------------------+---------+------+---------------+---------------------------------------------+
| id                       | estRows | task | access object | operator info                               |
+--------------------------+---------+------+---------------+---------------------------------------------+
| PartitionUnion_9         | 4.00    | root |               |                                             |
| ├─Batch_Point_Get_10     | 2.00    | root | table:t1      | handle:[1 10], keep order:false, desc:false |
| └─Batch_Point_Get_11     | 2.00    | root | table:t1      | handle:[1 10], keep order:false, desc:false |
+--------------------------+---------+------+---------------+---------------------------------------------+
3 rows in set (0.01 sec)
```

### Release note

- support BatchPointGet for partition table

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -691,7 +691,7 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty, planCounter
 				p: dual,
 			}, cntPlan, nil
 		}
-		canConvertPointGet := (!ds.isPartition && len(path.Ranges) > 0) || (ds.isPartition && len(path.Ranges) == 1)
+		canConvertPointGet := (!ds.isPartition || ds.isPartition) && len(path.Ranges) > 0
```
Comment (xiongjiwei) [inline planner/core/find_best_task.go:694]: `(!ds.isPartition || ds.isPartition)` always `true`

Comment (xiaodong-ji) [reply planner/core/find_best_task.go]: fixed

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -691,7 +691,7 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty, planCounter
 				p: dual,
 			}, cntPlan, nil
 		}
-		canConvertPointGet := (!ds.isPartition && len(path.Ranges) > 0) || (ds.isPartition && len(path.Ranges) == 1)
+		canConvertPointGet := len(path.Ranges) > 0
```
Comment (winoros) [inline planner/core/find_best_task.go:694]: @lysu For partition table, multi-range can also triger `Get`/`BatchGet` operation, right?

Comment (lysu) [reply planner/core/find_best_task.go:694]: I guess maybe some reason let https://github.com/pingcap/tidb/pull/14775#discussion_r388731250 only support single range? @lzmhhh123

Comment (lzmhhh123) [reply planner/core/find_best_task.go:694]: Sorry, I forgot the reason. Maybe the previous point get can't handle the partition table.

Comment (lysu) [reply planner/core/find_best_task.go:694]: oh...current point get can handle the partition

**Comment by Yisaer** [comment]
I think only modifying `planner` is not enough as I got the following result after I applied the changing:

``` following
mysql> create table t1 (c int primary key)
    -> PARTITION BY RANGE (c) (
    ->     PARTITION p0 VALUES LESS THAN (6),
    ->     PARTITION p1 VALUES LESS THAN (11),
    ->     PARTITION p2 VALUES LESS THAN (16),
    ->     PARTITION p3 VALUES LESS THAN (21)
    -> );
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1 (c) values (1);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t1 (c) values (0);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t1 (c) values (10);
Query OK, 1 row affected (0.00 sec)

mysql> explain select * from t1 where c in (1,10);
+--------------------------+---------+------+---------------+---------------------------------------------+
| id                       | estRows | task | access object | operator info                               |
+--------------------------+---------+------+---------------+---------------------------------------------+
| PartitionUnion_9         | 4.00    | root |               |                                             |
| ├─Batch_Point_Get_10     | 2.00    | root | table:t1      | handle:[1 10], keep order:false, desc:false |
| └─Batch_Point_Get_11     | 2.00    | root | table:t1      | handle:[1 10], keep order:false, desc:false |
+--------------------------+---------+------+---------------+---------------------------------------------+
3 rows in set (0.00 sec)

mysql> select * from t1 where c in (1,10);
Empty set (0.00 sec)

```

The `BatchPointGetExecutor` is also needed to be revised. @xiaodong-ji

**Comment by zz-jason** [comment]
@xiaodong-ji please add some tests to ensure the correctness and performance improvements, we need these tests to check:

- execution plan changes are correct and as expected
- query result is correct
- performance is improved

**Pull request closed without merging**